Sources/IntegerUtilities/Shift.swift

```diff
@@ -0,0 +1,149 @@
+//===--- Shift.swift ------------------------------------------*- swift -*-===//
+//
+// This source file is part of the Swift Numerics open source project
+//
+// Copyright (c) 2021 Apple Inc. and the Swift Numerics project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+extension BinaryInteger {
+  /// self / 2^(count), rounded the results according to the specified rule.
+  ///
+  /// The default rounding rule is `.down`, which matches the behavior of
+  /// the `>>` operator from the standard library.
+  ///
+  /// Some examples of different rounding rules:
+  ///
+  ///     // 3/2 is 1.5, which rounds (down by default) to 1.
+  ///     3.shifted(right: 1)
+  ///
+  ///     // 1.5 rounds up to 2.
+  ///     3.shifted(right: 1, rounding: .up)
+  ///
+  ///     // The two closest values are 1 and 2, 1 is returned because it
+  ///     // is odd.
+  ///     3.shifted(right: 1, rounding: .toOdd)
+  ///
+  ///     // 7/2^2 = 1.75, so the result is 1 with probability 1/4, and 2
+  ///     // with probability 3/4.
+  ///     7.shifted(right: 2, rounding: .stochastic)
+  ///
+  ///     // 4/2^2 = 4/4 = 1, exactly.
+  ///     4.shifted(right: 2, rounding: .trap)
+  ///
+  ///     // 5/2 is 2.5, which is not exact, so this traps.
+  ///     5.shifted(right: 1, rounding: .trap)
+  @inlinable
+  public func shifted<Count: BinaryInteger>(
+    right count: Count,
+    rounding rule: RoundingRule = .down
```

Comment:
As someone who doesn't think about rounding very often I'm just going to assume that you've thought about this and there's an obvious default here, but I'm leaving this comment just on the tiny chance that on seeing it again you'll go "hm actually maybe they should always have to say what they want" :)

Reply:
For right shift, the default should certainly be to truncate.  That's `down` or `towardsZero` for positive values.  I have to stop and think for a bit before I can remember the right behavior for negative values.

Reply:
Right; "normal" right shifts (>>) round down, and there's no good reason to make anything else the default. "Normal" division (/) rounds toward zero, which is why divide by power of two can't be replaced with a shift for signed values, but the way to resolve that mismatch is to make the forthcoming divide-with-rounding default to .down as well, as that gives a more sensible remainder for most uses.